server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java

```diff
@@ -49,6 +49,22 @@
      */
     InputStream readBlob(String blobName) throws IOException;
 
+    /**
+     * Creates a new {@link InputStream} that can be used to read the given blob starting from
+     * a specific {@code position} in the blob. The {@code length} is an indication of the
+     * number of bytes that are expected to be read from the {@link InputStream}.
+     *
+     * @param blobName The name of the blob to get an {@link InputStream} for.
+     * @param position The position in the blob where the next byte will be read.
+     * @param length   An indication of the number of bytes to be read.
+     * @return The {@code InputStream} to read the blob.
+     * @throws NoSuchFileException if the blob does not exist
+     * @throws IOException         if the blob can not be read.
+     */
+    default InputStream readBlob(final String blobName, final long position, final int length) throws IOException {
```

Comment:
Why would we want `length` on this API? Wouldn't it be better to just have `IndexInput` keep a reference to an open stream and only open a new stream if we seek backwards instead of opening a new stream of bounded length repeatedly?

Reply:
It could be done like you suggest for FS repository (and maybe HDFS too) but for other repositories we need to give an indication of the number of bytes we want to download, because unlike the RestoreService we don't want to read all the blobs but only a chunk of it. Most SDK require to consume all the requested bytes (or will consume them under the hood for you) and we don't want to open a stream that reads a complete blob if we only use the first 28 bytes to read a header.

Reply:
> Most SDK require to consume all the requested bytes (or will consume them under the hood for you) and we don't want to open a stream that reads a complete blob if we only use the first 28 bytes to read a header.

Fair point :) You have `abort` as a method on the S3 input stream though, not sure about GCS here.